### ISSUE TITLE
Seg fault while running GraphStorm using WholeGraph

### ISSUE THREAD

**Issue by isratnisa** [rapidsai/wholegraph#85]
We are using WholeGraph in [GraphStorm](https://github.com/awslabs/graphstorm) to train GNN models. We encountered `Segmentation fault` when using WholeGraph with 8 trainers on 4 nodes. 4 trainers on 4 nodes rarely throws this error but usually clearing shared memory helps. Is it an OOM issue for 8 trainers? Issue couldn’t be reproduced for baseline (w/o wholegraph).

```
[ip-172-31-20-63:49779:0:49779] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x7f96ef392140)
==== backtrace (tid:  49779) ====
 0 0x0000000000042520 __sigaction()  ???:0
 1 0x000000000002a2e8 __pyx_f_15pylibwholegraph_7binding_19wholememory_binding_python_cb_wrapper_temp_free()  wholememory_binding.cxx:0
 2 0x000000000027428c wholememory_ops::wholememory_gather_nccl()  ???:0
 3 0x0000000000272b75 wholememory_gather()  ???:0
 4 0x00000000000e9225 wholememory::noncached_embedding::gather()  ???:0
 5 0x0000000000056fa2 __pyx_pw_15pylibwholegraph_7binding_19wholememory_binding_13EmbeddingGatherForward()  wholememory_binding.cxx:0
 6 0x000000000015372b _PyObject_MakeTpCall()  ???:0
 7 0x000000000014c0e7 _PyEval_EvalFrameDefault()  ???:0
 8 0x000000000015d4ec _PyFunction_Vectorcall()  ???:0
 9 0x0000000000145c14 _PyEval_EvalFrameDefault()  ???:0
10 0x000000000015d4ec _PyFunction_Vectorcall()  ???:0
11 0x0000000000146d6b _PyEval_EvalFrameDefault()  ???:0
12 0x000000000015d4ec _PyFunction_Vectorcall()  ???:0
13 0x0000000000145c14 _PyEval_EvalFrameDefault()  ???:0
14 0x000000000016af11 PyMethod_New()  ???:0
15 0x0000000000146d6b _PyEval_EvalFrameDefault()  ???:0
16 0x000000000015d4ec _PyFunction_Vectorcall()  ???:0
17 0x0000000000145a1d _PyEval_EvalFrameDefault()  ???:0
18 0x0000000000142176 _PyArg_ParseTuple_SizeT()  ???:0
19 0x0000000000237c56 PyEval_EvalCode()  ???:0
20 0x0000000000264b18 PyUnicode_Tailmatch()  ???:0
21 0x000000000025d96b PyInit__collections()  ???:0
22 0x0000000000264865 PyUnicode_Tailmatch()  ???:0
23 0x0000000000263d48 _PyRun_SimpleFileObject()  ???:0
24 0x0000000000263a43 _PyRun_AnyFileObject()  ???:0
25 0x0000000000254c3e Py_RunMain()  ???:0
26 0x000000000022abcd Py_BytesMain()  ???:0
27 0x0000000000029d90 __libc_init_first()  ???:0
28 0x0000000000029e40 __libc_start_main()  ???:0
29 0x000000000022aac5 _start()  ???:0
=================================
[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49778 closing signal SIGTERM
[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49780 closing signal SIGTERM
[[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49781 closing signal SIGTERM
14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:390: 
User pressed Ctrl+C, Exiting
[[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49782 closing signal SIGTERM
14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:390: 
User pressed Ctrl+C, Exiting
[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49783 closing signal SIGTERM
[14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:[[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49784 closing signal SIGTERM
390: 
User pressed Ctrl+C, Exiting
14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:390[: 
User pressed Ctrl+C, Exiting
[2023-10-20 14:30:07,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Sending process 49785 closing signal SIGTERM
14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:390: 
User pressed Ctrl+C, Exiting
[14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:390: 
User pressed Ctrl+C, Exiting
[14:30:07] /opt/dgl/dgl-source/src/rpc/rpc.cc:390: 
User pressed Ctrl+C, Exiting
terminate called without an active exception
terminate called without an active exception
terminate called without an active exception
terminate called without an active exception
terminate called without an active exception
terminate called without an active exception
terminate called without an active exception
[2023-10-20 14:30:37,147] torch.distributed.elastic.multiprocessing.api: [WARNING] Unable to shutdown process 49778 via Signals.SIGTERM, forcefully exiting via Signals.SIGKILL
[2023-10-20 14:30:41,255] torch.distributed.elastic.multiprocessing.api: [WARNING] Unable to shutdown process 49780 via Signals.SIGTERM, forcefully exiting via Signals.SIGKILL
[2023-10-20 14:30:48,381] torch.distributed.elastic.multiprocessing.api: [WARNING] Unable to shutdown process 49783 via Signals.SIGTERM, forcefully exiting via Signals.SIGKILL
[2023-10-20 14:30:53,370] torch.distributed.elastic.multiprocessing.api: [ERROR] failed (exitcode: -11) local_rank: 1 (pid: 49779) of binary: /usr/bin/python3
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launch.py", line 196, in <module>
    main()
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launch.py", line 192, in main
    launch(args)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launch.py", line 177, in launch
    run(args)
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/run.py", line 788, in run
    elastic_launch(
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 134, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/usr/local/lib/python3.10/dist-packages/torch/distributed/launcher/api.py", line 264, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
/workspace/graphstorm/python/graphstorm/run/gsgnn_np/gsgnn_np.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2023-10-20_14:30:07
  host      : ip-172-31-20-63.us-west-2.compute.internal
  rank      : 1 (local_rank: 1)
  exitcode  : -11 (pid: 49779)
  error_file: <N/A>
  traceback : Signal 11 (SIGSEGV) received by PID 49779
============================================================
```

**Comment by nvcastet** [comment]
@dongxuy04 Do you have an idea on what could be the issue by looking at the stack trace?

**Comment by dongxuy04** [comment]
@isratnisa I am not sure if it is OOM issue for 8 trainers based on this, I think monitor output of CPU/GPU memory usage and dmesg output may help on this. BTW, as you mentioned 'clearing shared memory helps', is distributed host memory used in this case? And how about the machine config

**Comment by isratnisa** [comment]
Yes, it's a distributed memory setup. We are using two p4de EC2 instances, each of which is equipped with eight A100 GPUs, with each GPU having 80GB of memory. The issue is more frequent when we use 8 trainers which utilizes all 8 GPUs instead of 4 trainers.

You are right! Memory might not be the issue. We are only using ~2GB of each GPU and also `free -m` command shows that we have free memory.

**Comment by nvcastet** [comment]
I also got a similar stacktrace:
```
 3: [eos0055:3059617:0:3059617] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x7ffdc0152680)
 3: ==== backtrace (tid:3059617) ====
 3:  0 0x0000000000042520 __sigaction()  ???:0
 3:  1 0x000000000004912c __pyx_f_15pylibwholegraph_7binding_19wholememory_binding_python_cb_wrapper_temp_free()  wholememory_binding.cxx:0
 3:  2 0x00000000002d671e wholememory_ops::wm_thrust_allocator::deallocate()  ???:0
 3:  3 0x00000000001cf7c6 wholememory_ops::exchange_ids_temp_func<long>()  ???:0
 3:  4 0x00000000001c65c5 wholememory_ops::exchange_ids_func()  ???:0
 3:  5 0x00000000001c70a7 wholememory_ops::bucket_and_exchange_ids_func()  ???:0
 3:  6 0x00000000002b1ad5 wholememory_ops::wholememory_gather_nccl()  ???:0
 3:  7 0x00000000002b0e7e wholememory_gather()  ???:0
 3:  8 0x0000000000134f75 wholememory::noncached_embedding::gather()  ???:0
 3:  9 0x000000000005dc26 __pyx_pw_15pylibwholegraph_7binding_19wholememory_binding_13EmbeddingGatherForward()  wholememory_binding.cxx:0
 3: 10 0x000000000014e8a2 _PyEval_EvalFrameDefault()  ???:0
 3: 11 0x000000000016070c _PyFunction_Vectorcall()  ???:0
 3: 12 0x0000000000148f52 _PyEval_EvalFrameDefault()  ???:0
 3: 13 0x000000000016070c _PyFunction_Vectorcall()  ???:0
 3: 14 0x000000000014a0d1 _PyEval_EvalFrameDefault()  ???:0
 3: 15 0x000000000016070c _PyFunction_Vectorcall()  ???:0
 3: 16 0x0000000000148f52 _PyEval_EvalFrameDefault()  ???:0
 3: 17 0x000000000016e4e1 PyMethod_New()  ???:0
 3: 18 0x000000000014a0d1 _PyEval_EvalFrameDefault()  ???:0
 3: 19 0x000000000016070c _PyFunction_Vectorcall()  ???:0
 3: 20 0x0000000000148e0d _PyEval_EvalFrameDefault()  ???:0
 3: 21 0x0000000000239e56 PyEval_EvalCode()  ???:0
 3: 22 0x0000000000239cf6 PyEval_EvalCode()  ???:0
 3: 23 0x00000000002647d8 PyUnicode_Tailmatch()  ???:0
 3: 24 0x000000000025e0bb PyInit__collections()  ???:0
 3: 25 0x0000000000264525 PyUnicode_Tailmatch()  ???:0
 3: 26 0x0000000000263a08 _PyRun_SimpleFileObject()  ???:0
 3: 27 0x0000000000263653 _PyRun_AnyFileObject()  ???:0
 3: 28 0x000000000025641e Py_RunMain()  ???:0
 3: 29 0x000000000022ccad Py_BytesMain()  ???:0
 3: 30 0x0000000000029d90 __libc_init_first()  ???:0
 3: 31 0x0000000000029e40 __libc_start_main()  ???:0
 3: 32 0x000000000022cba5 _start()  ???
 ```
 
The stacktrace is slightly different but still seems to crash in `wholememory_binding_python_cb_wrapper_temp_free() `.
@dongxuy04 Do you know how could we debug that issue? For my side it is not super reprocible, but i think @isratnisa can reproduce it easily.

**Comment by dongxuy04** [comment]
@nvcastet This stack seems strange to me. It seems that this happens in `python_cb_wrapper_temp_free`, it is dealing with some Python objects and call Python Callback function. Maybe can add some logs in `wholememory_binding.pyx` to check if the Python objects are valid and refcnt is positive.

**Comment by linhu-nv** [comment]
Hi @isratnisa @nvcastet, I got a similar bug from other users a few days ago. They were using wholegraph on 16 nodes X 8 ranks distributed environments, and their bug had similar backtraces. It turns out that it's because NCCL 2.19.x has a bug, and add  `NCCL_CUMEM_ENABLE=0` could fix it. This bug is also fixed in NCCL 2.20.x. Wish this could help. Please let me know if you have any other questions.
